Engine/DataFeeds/WorkScheduling/WorkItem.cs

```diff
@@ -0,0 +1,79 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+using System;
+
+namespace QuantConnect.Lean.Engine.DataFeeds.WorkScheduling
+{
+    internal class WorkItem
+    {
+        private readonly Func<int> _weightFunc;
+
+        /// <summary>
+        /// The current weight
+        /// </summary>
+        public int Weight { get; private set; }
+
+        /// <summary>
+        /// The work function to execute
+        /// </summary>
+        public Func<int, bool> Work { get; }
```

Comment:
Since these are being executed a bagillion times, it might show an improvement to make this a class method. Could makeit accept `Expression<Func<int, bool>>` (simplify each func to just a static method call) and then compile them to `Func<int,bool>` and invoke from sealed class methods, and _actually_ this `WorkItem` will likely perform better as a `struct` -- bah -- struct would likely be dangerous given out the update weights work...hmm... the weight could be replaced by a `Ref<int>` as the weight, keeping the struct data immutable but adding a little bit to the access time. Anyway, another benchmark that I think would be interesting. 

The theory behind making it a struct is to remove the heap allocations, especially for things that are happening millions/billions of times. another solution to removing heap allocations is object pooling. You make the object mutable and you reuse them. IIRC I wrote an object pool in lean or perhaps another repo, but either way,they're pretty easy to write.